### PR TITLE
Minor improvements to ECC arithmetic subroutines

### PR DESCRIPTION
**Pull request by hanno-arm**
This PR makes a number of small improvements to ECC arithmetic subroutines:
* https://github.com/ARMmbed/mbedtls/pull/5377/commits/5c8ea307b878a9860996972c5890c53d8655c525, https://github.com/ARMmbed/mbedtls/pull/5377/commits/76f897d6994ca95c64edf527127ecb9b82a0d071, https://github.com/ARMmbed/mbedtls/pull/5377/commits/0d629791e9143056d9b2c2eca793a02256054211, https://github.com/ARMmbed/mbedtls/pull/5377/commits/28ccb1cc90101bed8befcdfbdb961b5413975a2e, https://github.com/ARMmbed/mbedtls/pull/5377/commits/02a999b91a02144ba34ca9352d297408d37c67e8 and https://github.com/ARMmbed/mbedtls/pull/5377/commits/b8442cd9c6323a6a8255eecc94617becee5056ae remove local MPIs from ECP arithmetic subroutines
* https://github.com/ARMmbed/mbedtls/pull/5377/commits/ce29ae84dd425877178b81ee9fdf1491ef623943,  https://github.com/ARMmbed/mbedtls/pull/5377/commits/02b35bd00a0b2c100975b0b0575a74ec5250324b and https://github.com/ARMmbed/mbedtls/pull/5377/commits/885ed403c954eebe5c16abec2c6bc99091073193 improve readability of ECP code by introducing macros for low-level modular arithmetic.
* https://github.com/ARMmbed/mbedtls/pull/5377/commits/a7f8edd709e7df569b22ce5f5a3648669e577699 and https://github.com/ARMmbed/mbedtls/pull/5377/commits/3b29f2194b924acec282d31d19c8c15da22b7814 reduce heap allocations by using temporary MPIs across iterated calls to the same ECC arithmetic routines

Edit: avoid merging until we're confident #5383 will also be merged before the next release.

### CODE REVIEW

File: library/ecp.c
Comment (mpg) [inline library/ecp.c:1476]: Minor: I'm slightly concerned that the aliases might hide the fact that we're writing to the output, hence potentially overwriting the input. Perhaps leaving `&R->X` and so on in the code would make that more apparent? Either ways, we should leave a comment about it.

File: library/ecp.c
Comment (mpg) [inline library/ecp.c:1387]: Note: I seemed to remember we implemented multiplication by 3 as "left-shit by 1 then add" but apparently not. Not sure if it makes a difference, actually. (No change required or even suggested, just thinking out loud.)

**Review by mpg — APPROVED** [review]
Looks good to me. Just a slight maintainability concern, which is not a blocker as I'm not sure it's really a problem (if we ever modify this code again, we're very likely to pay attention to that issue anyway).

File: library/ecp.c
Comment (hanno-arm) [reply library/ecp.c:1476]: Agree

File: library/ecp.c
Comment (hanno-arm) [reply library/ecp.c:1387]: Yes, good point... we should experiment with at some point.

**Comment by hanno-arm** [comment]
@mpg Thanks for your review, I believe to have addressed your comments, and also added two more commits improving readibility and removing two local MPIs from `ecp_normalize_jac_many()`. Could you take another look?

**Review by mpg — APPROVED** [review]
LGTM so far, as long as the CI agrees. (I think this is an intermediate state as the preparation for keeping temporaries with mxz was done, but the final change not executed yet. However I think it's expected that this PR is not final yet.)

**Comment by mpg** [comment]
Related: https://github.com/ARMmbed/mbedtls/pull/5381 - I ran `ecc-heap.sh` (with trivial editing to restrict it to interesting results) and got the following results:

```
==> before <==
fixed point optim = 1, max window size = 4
--------------------------------------------

  ECDHE-secp256r1          :      86 full handshake/s   4720 heap bytes,  62454 allocs
  ECDHE-x25519             :     282 full handshake/s   2456 heap bytes,  15476 allocs

==> after <==
fixed point optim = 1, max window size = 4
--------------------------------------------

  ECDHE-secp256r1          :     110 full handshake/s   5216 heap bytes,  57468 allocs
  ECDHE-x25519             :     305 full handshake/s   2432 heap bytes,  18542 allocs
```
The number of allocs decreased for P-256, though not much so far, and surprisingly the peak usage increased. This might be due to more of the temporaries ending up double-sized, since there are fewer and they're longer-lived. I'm looking forward to the results with a single double-sized temporary. (For x25519, there's no reduction in number of allocs so far, probably because the temporaries are not long-lived yet.)

**Comment by hanno-arm** [comment]
@mpg Thank you! I suspect that the larger heap usage is due to the fact that the output in https://github.com/ARMmbed/mbedtls/commit/5c8ea307b878a9860996972c5890c53d8655c525 is now double-width, where it was previously single-width. This is only temporary, because we intend to use a single double-width temporary for the intermediate result.

**Comment by hanno-arm** [comment]
@mpg I suspect I'm doing something stupid, but I can't reproduce your numbers. For this PR, I'm getting

```
ECDHE-secp256r1          :     456 full handshake/s   5696 heap bytes,   3028 allocs
```

while for https://github.com/ARMmbed/mbedtls/pull/5381 I get

```
 ECDHE-secp256r1          :     397 full handshake/s   4496 heap bytes,  10892 allocs
```

**Comment by mpg** [comment]
For the record: the discrepancy in numbers was because Hanno was using the default config plus `MEMORY_DEBUG` (and its dependencies), while `ecc-heap.sh` is using a standalone reduced config. It turns out this makes a large difference because of `MBEDTLS_ECP_NIST_OPTIM`: without it, reduction modulo P after multiplication is done using `mbedtls_mpi_mod_mpi()` rather than a curve-specific optimized routine, resulting in many more allocations.

I updated 5381 to enable `NIST_OPTIM` in order to get more reasonable figures.

**Comment by mpg** [comment]
Note: output from the new #5381 extended with PSA as suggested by Gilles:
```
fixed point optim = 1, max window size = 4
--------------------------------------------

  ECDSA-secp256r1          :    2340 sign/s             1624 heap bytes,   1536 allocs
  ECDSA-secp256r1          :     698 verify/s           3912 heap bytes,   5616 allocs
  ECDHE-secp256r1          :     368 full handshake/s   4496 heap bytes,  10948 allocs
  ECDHE-x25519             :     259 full handshake/s   2456 heap bytes,  15478 allocs
```
ECDSA-sign is (except for some arithmetic mod n) basically a fixed-point multiplication, that is a multiplication where `ecp_precompute_comb()` is skipped because the pre-computation was done off-line and the results embedded in the library.

ECDSA-verify is (except for one point addition) basically a fixed-point multiplication plus a variable-point multiplication (that is, one where `ecp_precompute_comb()` is not skipped). That suggests that a variable-point multiplication is about 4100 allocs, of which about 1500 would be outside `ecp_precompute_comb()`, hence the remaining 2600 in `ecp_precompute_comb()`. That means it's probably worth investigating how to optimize that part of the computation.

So far work to re-use temporaries has mostly touched `ecp_mul_comb_core()` which is the part that happens after pre-computation (it's called from `ecp_mul_comb_after_precomp()`).

**Comment by hanno-arm** [comment]
@mpg Could you review the latest commit? 

How would you suggest to proceed? On the one hand, I would like to avoid accumulating overly large PRs; on the other hand, without the follow-up PR #5383, this PR increases heap usage non-negligibly.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
Correctness: looks mostly good to me, but there are two places where I'm not completely sure:
    * [about aliasing]()
    * [that calling lset unconditionally in `ecp_select_comb` is always correct]()

Maintenability: looks mostly good to me, but some comments need updating. I like that we're on a path where ECP can become more independent of bignum.

Performance: I'm very happy that the number of allocations is going down. I haven't run any benchmarks. Have you observed a difference in speed in some builds?

As you note, what really matters is the effect of this plus #5383. Despite the heap usage increase, I'm ok with merging this PR now if we're reasonably confident that #5383 is likely to be merged before the next release. This PR is very disruptive in terms of code changes, so it will bitrot fast if we make any changes to `ecp.c`, and therefore I don't want it to linger.

File: library/ecp.c
Comment (hanno-arm) [reply library/ecp.c]: Yes, I didn't think about finer grained aliasing than on the level of points. I don't think we need to worry about this, but will leave a comment at the top of the function that aliasing is supported at the level of points.

File: library/ecp.c
Comment (hanno-arm) [reply library/ecp.c:1346]: Will do.

File: library/ecp.c
Comment (hanno-arm) [reply library/ecp.c:1659]: Ok.

File: library/ecp.c
Comment (hanno-arm) [reply library/ecp.c:2469]: Thank you for working through this, I'm happy to integrate your comments.

File: library/ecp.c
Comment (hanno-arm) [reply library/ecp.c:1460]: Right, I'll fix that.

File: library/ecp.c
Comment (hanno-arm) [reply library/ecp.c]: Great spot, you're right. Except for the storage-form of the `T[]`-array, which however is only ever indirectly accessed through `ecp_select_comb()` which does also set the Z coordinate, we can get rid of 'incomplete' EC points altogether now.

File: library/ecp.c
Comment (hanno-arm) [reply library/ecp.c:1652]: Fixing.

**Comment by hanno-arm** [comment]
Thanks @gilles-peskine-arm for your review. I believe to have addressed your comments.

**Comment by hanno-arm** [comment]
```
CC=arm-none-eabi-gcc CFLAGS="-Os -mthumb -mcpu=cortex-m0plus" ./scripts/code_size_compare.py -o development -n 466df6e713fb743a7960f45774ba8470a6ef4504 -s -c crypto_baremetal -t arm_none_eabi_gcc_8_4_3_baremetal_Os_M0plus
```
reports 1% code size reduction.

**Comment by mpg** [comment]
I'm applying "needs: preceding PR" even though it doesn't strictly apply, because we don't have a "needs: follow-up PR in the same release" label and that's the closest approximation.

**Review by gilles-peskine-arm — APPROVED** [review]
Looks good to me.

Regarding whether to wait for follow-ups: looking again at the numbers, I don't find them dramatic. Peak heap usage has increased, and this can be the top heap usage in a plausible application, but it hasn't increased so much that I find it unacceptable, and on the flip side fewer allocations means that heap fragmentation might be going down so the usable heap size might not be that much more. Code size has only very slightly increased. So I now think we should merge this PR as soon as it's approved and not risk bitrotting.

**Comment by hanno-arm** [comment]
@mpg Do you think you have time to re-review this?

**Comment by mpg** [comment]
It's on my list, and baring surprises, I should be able to get to it next week.

**Comment by hanno-arm** [comment]
Any update, @mpg ?

**Comment by mpg** [comment]
The things that were before this PR on my list are taking a bit longer than expected to converge. I should be able to get to this PR this week.

**Review by mpg — APPROVED** [review]
LGTM

**Comment by mpg** [comment]
@hanno-arm @gilles-peskine-arm What do we want to do with this? We can either:

1. Merge it now, and hope we can review and merge #5383 before the next release (or accept a 25% increase in max heap usage for some curves).
2. Wait until 5383 is also reviewed.

Currently I'm more inclined to merge now. While we don't have a target date for the next release, it's likely to be end of March or later, so there's a good chance we can review and merge 5383 by then; and in case we don't, the consequences are not that dramatic. Wdyt?

**Comment by gilles-peskine-arm** [comment]
After discussing with the team: we're merging this right now. We'll try to get the follow-ups in before the next release. We'll create an epic to track that.

**Pull request merged**
